keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -0,0 +1,766 @@
+
+```
+cosi.io/provisioner-secret-name:
+cosi.io/provisioner-secret-namespace:
+```
+
+- **Per BucketClass:** A secret may be made specific to a BucketClass.  This suits cases where authorization may be segregated in the object store.  The Secret may then be defined explicitly in the `bucketClass.parameters` map.
+
+  ```
+  cosi.io/provisioner-secret-name: "foo"
+  cosi.io/provisioner-secret-namespace: "bar"
+  ```
+
+- **Per Operation/Bucket:** Unique credentials are passed per Bucket or operation. In order to support dynamic Secret naming, templating similar to [CSI Secret templating](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html) may be used.  E.g.
+
+  ```
+  "${bucket.name}"
+  "${bucket.namespace}"
+  ```
+  
+  Admins may then define a BucketClass with the following parameters included:
+
+  *Per Bucket Operation*
+  
+  ```
+  cosi.io/provisioner-secret-name: "${bucket.name}"
+  cosi.io/provisioner-secret-namespace: "${bucket.namespace}"
+  ```
+---
+
+# gRPC Definitions
+
+There is one service defined by COSI - `Provisioner`. This will need to be satisfied by the vendor-provisioner in order to be COSI-compatible
+
+```
+service Provisioner {
+    rpc ProvisionerGetInfo (ProvisionerGetInfoRequest) returns (ProvisionerGetInfoResponse) {}
+
+    rpc ProvisionerCreateBucket (ProvisionerCreateBucketRequest) returns (ProvisionerCreateBucketResponse) {}
+    rpc ProvisionerDeleteBucket (ProvisionerDeleteBucketRequest) returns (ProvisionerDeleteBucketResponse) {}
+
+    rpc ProvisionerGrantBucketAccess (ProvisionerGrantBucketAccessRequest) returns (ProvisionerGrantBucketAccessResponse);
+    rpc ProvisionerRevokeBucketAccess (ProvisionerRevokeBucketAccessRequest) returns (ProvisionerRevokeBucketAccessResponse);
+}
+```
+
+## ProvisionerGetInfo
+
+This call is meant to retrieve the unique provisioner Identity. This identity will have to be set in `BucketRequest.Provisioner` field in order to invoke this specific provisioner.
+
+```
+message ProvisionerGetInfoRequest {
+    // Intentionally left blank
+}
+
+message ProvisionerGetInfoResponse {    
+    string provisioner_identity = 1;
+}
+```
+
+## ProvisonerCreateBucket
+
+This call is made to create the bucket in the backend. If the bucket already exists, then the appropriate error code `ALREADY_EXISTS` must be returned by the provisioner.
+
+```
+message ProvisionerCreateBucketRequest {    
+    // This field is REQUIRED
+    string bucket_name = 1;
+
+    map<string,string> bucket_context = 2;
+
+    enum AnonymousBucketAccessMode {
+	PRIVATE = 0;
+	PUBLIC_READ_ONLY = 1;
+	PUBLIC_WRITE_ONLY = 2;
+	PUBLIC_READ_WRITE = 3;
+    }
+    
+    AnonymousBucketAccessMode anonymous_bucket_access_mode = 3;
+}
+
+message ProvisionerCreateBucketResponse {
+    // Intentionally left blank
+}
```

Comment:
Why is this intentionally left blank?  Shouldn't this be returning a bucket object so you can check whether bucket is created successfully or not?  In the CSI model, the response would include a resource which contains an ID generated by the storage system.

Reply:
@wlan0 @brahmaroutu

Reply:
In CSI, the identity of the FS or Block device (device UUID) does not matter for the pod. In case of object storage, the bucket name matters. Therefore, COSI controls the ID of the generated resource. It is not a part of the response. 

gRPC provides a separate channel for error scenarios. We utilize that for returning errors if any.

Reply:
How do you know whether a bucket is provisioned or not?  CreateBucket should be idempotent.  How do you support that?

Reply:
Is there a GetBucket?  Does CreateBucket also work as GetBucket?

Reply:
So CreateBucket in the backendAPI in Idempotent. I.e. for a given bucket name, two calls to MakeBucket will end up creating only one bucket